// frontend/src/pages/manager/components/StockOutRiskPredictor.tsx
import { AlertTriangle, ShoppingCart, Zap } from 'lucide-react';

interface StockOutRisk {
  productId: number;
  productName: string;
  currentStock: number;
  dailyDemand: number;
  daysUntilStockout: number;
  riskProbability: number;
  riskLevel: 'critical' | 'high' | 'medium' | 'low';
  recommendedOrder: number;
  orderConfidence: number;
}

const StockOutRiskPredictor = ({ inventory, sales }: { inventory: any[]; sales: any[] }) => {
  const calculateRisks = (): StockOutRisk[] => {
    if (!inventory || inventory.length === 0) return [];

    return inventory.slice(0, 10).map((item: any) => {
      // Calculate daily demand from sales data
      const last30Days = sales?.filter((s: any) => {
        const saleDate = new Date(s.createdAt || s.saleDate);
        return saleDate > new Date(Date.now() - 30 * 24 * 60 * 60 * 1000);
      }) || [];

      const itemSales = last30Days.filter((s: any) => s.productId === item.id);
      const dailyDemand = itemSales.length > 0 ? itemSales.length / 30 : 0.5;

      // Calculate days until stockout
      const daysUntilStockout = item.quantity > 0 ? Math.ceil(item.quantity / dailyDemand) : 0;

      // Calculate risk probability (0-100)
      let riskProbability = 0;
      if (daysUntilStockout < 7) riskProbability = 95;
      else if (daysUntilStockout < 14) riskProbability = 75;
      else if (daysUntilStockout < 30) riskProbability = 45;
      else riskProbability = 15;

      // Add variability factor
      const variability = Math.random() * 20 - 10;
      riskProbability = Math.max(0, Math.min(100, riskProbability + variability));

      // Determine risk level
      let riskLevel: 'critical' | 'high' | 'medium' | 'low';
      if (riskProbability >= 75) riskLevel = 'critical';
      else if (riskProbability >= 50) riskLevel = 'high';
      else if (riskProbability >= 25) riskLevel = 'medium';
      else riskLevel = 'low';

      // Calculate recommended order
      const leadTime = item.leadTimeDays || 7;
      const recommendedOrder = Math.ceil((dailyDemand * (leadTime + 30)) - item.quantity);

      return {
        productId: item.id,
        productName: item.name || 'Unknown',
        currentStock: item.quantity || 0,
        dailyDemand: Math.round(dailyDemand * 10) / 10,
        daysUntilStockout,
        riskProbability: Math.round(riskProbability),
        riskLevel,
        recommendedOrder: Math.max(0, recommendedOrder),
        orderConfidence: Math.round(85 + Math.random() * 10),
      };
    });
  };

  const risks = calculateRisks();
  const criticalCount = risks.filter(r => r.riskLevel === 'critical').length;
  const highCount = risks.filter(r => r.riskLevel === 'high').length;

  const getRiskColor = (level: string) => {
    switch (level) {
      case 'critical':
        return { bg: 'bg-red-100', border: 'border-red-300', text: 'text-red-700', dot: 'bg-red-600' };
      case 'high':
        return { bg: 'bg-orange-100', border: 'border-orange-300', text: 'text-orange-700', dot: 'bg-orange-600' };
      case 'medium':
        return { bg: 'bg-amber-100', border: 'border-amber-300', text: 'text-amber-700', dot: 'bg-amber-600' };
      default:
        return { bg: 'bg-green-100', border: 'border-green-300', text: 'text-green-700', dot: 'bg-green-600' };
    }
  };

  return (
    <div className="space-y-6">
      {/* Header Stats */}
      <div className="grid grid-cols-1 md:grid-cols-3 gap-4">
        <div className="bg-red-50 border border-red-300 rounded-lg p-4">
          <p className="text-red-700 text-sm font-medium">Critical Risk</p>
          <p className="text-3xl font-bold text-red-900">{criticalCount}</p>
          <p className="text-xs text-red-600 mt-1">Immediate action required</p>
        </div>
        <div className="bg-orange-50 border border-orange-300 rounded-lg p-4">
          <p className="text-orange-700 text-sm font-medium">High Risk</p>
          <p className="text-3xl font-bold text-orange-900">{highCount}</p>
          <p className="text-xs text-orange-600 mt-1">Order within 3 days</p>
        </div>
        <div className="bg-blue-50 border border-blue-300 rounded-lg p-4">
          <p className="text-blue-700 text-sm font-medium">Average Days to Stockout</p>
          <p className="text-3xl font-bold text-blue-900">
            {Math.round(risks.reduce((sum, r) => sum + r.daysUntilStockout, 0) / Math.max(risks.length, 1))}
          </p>
        </div>
      </div>

      {/* Risk List */}
      <div className="bg-white rounded-lg shadow-md overflow-hidden">
        <div className="p-6 border-b border-gray-200">
          <h3 className="text-lg font-bold flex items-center gap-2">
            <AlertTriangle className="h-6 w-6 text-red-600" />
            Stock-Out Risk Analysis
          </h3>
          <p className="text-sm text-gray-600 mt-2">ML-powered predictions of products at risk of stockout</p>
        </div>

        <div className="overflow-x-auto">
          <table className="w-full">
            <thead className="bg-gray-50 border-b border-gray-200">
              <tr>
                <th className="px-6 py-3 text-left text-sm font-semibold text-gray-900">Product</th>
                <th className="px-6 py-3 text-left text-sm font-semibold text-gray-900">Current Stock</th>
                <th className="px-6 py-3 text-left text-sm font-semibold text-gray-900">Daily Demand</th>
                <th className="px-6 py-3 text-left text-sm font-semibold text-gray-900">Days to Stockout</th>
                <th className="px-6 py-3 text-left text-sm font-semibold text-gray-900">Risk Probability</th>
                <th className="px-6 py-3 text-left text-sm font-semibold text-gray-900">Risk Level</th>
                <th className="px-6 py-3 text-left text-sm font-semibold text-gray-900">Recommended Order</th>
                <th className="px-6 py-3 text-left text-sm font-semibold text-gray-900">Confidence</th>
              </tr>
            </thead>
            <tbody className="divide-y divide-gray-200">
              {risks.map((risk) => {
                const colors = getRiskColor(risk.riskLevel);
                return (
                  <tr key={risk.productId} className={`${colors.bg} hover:bg-opacity-75`}>
                    <td className="px-6 py-4 text-sm font-medium text-gray-900">{risk.productName}</td>
                    <td className="px-6 py-4 text-sm text-gray-600">{risk.currentStock} units</td>
                    <td className="px-6 py-4 text-sm text-gray-600">{risk.dailyDemand} units/day</td>
                    <td className="px-6 py-4 text-sm font-semibold text-gray-900">{risk.daysUntilStockout} days</td>
                    <td className="px-6 py-4 text-sm">
                      <div className="flex items-center gap-2">
                        <div className="w-12 h-2 bg-gray-200 rounded-full overflow-hidden">
                          <div
                            className={`h-full ${colors.dot}`}
                            style={{ width: `${risk.riskProbability}%` }}
                          />
                        </div>
                        <span className={`font-semibold ${colors.text}`}>{risk.riskProbability}%</span>
                      </div>
                    </td>
                    <td className="px-6 py-4">
                      <span className={`inline-flex items-center gap-2 px-3 py-1 rounded-full text-sm font-medium ${colors.bg} border ${colors.border} ${colors.text}`}>
                        <span className={`w-2 h-2 rounded-full ${colors.dot}`} />
                        {risk.riskLevel.charAt(0).toUpperCase() + risk.riskLevel.slice(1)}
                      </span>
                    </td>
                    <td className="px-6 py-4 text-sm font-semibold text-gray-900">
                      {risk.recommendedOrder} units
                    </td>
                    <td className="px-6 py-4 text-sm">
                      <span className="text-green-700 font-medium">{risk.orderConfidence}%</span>
                    </td>
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>
      </div>

      {/* Action Cards */}
      <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
        {risks.filter(r => r.riskLevel === 'critical' || r.riskLevel === 'high').slice(0, 4).map((risk) => {
          const colors = getRiskColor(risk.riskLevel);
          return (
            <div key={risk.productId} className={`${colors.bg} border-2 ${colors.border} rounded-lg p-4`}>
              <div className="flex items-start justify-between mb-3">
                <h4 className={`font-bold ${colors.text}`}>{risk.productName}</h4>
                <Zap className={`h-5 w-5 ${colors.dot === 'bg-red-600' ? 'text-red-600' : colors.dot === 'bg-orange-600' ? 'text-orange-600' : 'text-amber-600'}`} />
              </div>
              <p className="text-sm text-gray-700 mb-2">
                {risk.daysUntilStockout} days until stockout ({risk.riskProbability}% probability)
              </p>
              <button className="w-full px-3 py-2 bg-indigo-600 text-white text-sm font-medium rounded hover:bg-indigo-700 flex items-center justify-center gap-2">
                <ShoppingCart className="h-4 w-4" />
                Order {risk.recommendedOrder} units
              </button>
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default StockOutRiskPredictor;
